inc/cppxaml/XamlWindow.h

```diff
@@ -0,0 +1,256 @@
+                    auto xw = reinterpret_cast<XamlWindow*>(GetWindowLongPtr(hWnd, GWLP_USERDATA));
+                    if (xw == nullptr && message == WM_CREATE) {
+                        auto createStruct = reinterpret_cast<LPCREATESTRUCT>(lParam);
+
+                        xw = reinterpret_cast<XamlWindow*>(createStruct->lpCreateParams);
+                        SetWindowLongPtr(hWnd, GWLP_USERDATA, reinterpret_cast<LONG_PTR>(xw));
+                    }
+
+
+                    return xw ? xw->WndProc(hWnd, message, wParam, lParam) : DefWindowProc(hWnd, message, wParam, lParam);
+                };
+
+                wcex.cbClsExtra = 0;
+                wcex.cbWndExtra = sizeof(XamlWindow*);
+                wcex.hInstance = ctl->HInstance();
+                wcex.hCursor = LoadCursor(nullptr, IDC_ARROW);
+                wcex.hbrBackground = 0;
+                wcex.lpszMenuName = nullptr;
+                wcex.lpszClassName = WindowClass();
+                winrt::check_bool(RegisterClassExW(&wcex) != 0);
+            });
+
+            desktopXamlSource = winrt::Windows::UI::Xaml::Hosting::DesktopWindowXamlSource();
+            auto hWnd = CreateWindowW(WindowClass(), szTitle, style,
+                CW_USEDEFAULT, CW_USEDEFAULT, width, height, parent, nullptr, controller->HInstance(), this);
+            if (hWnd) {
+                ShowWindow(hWnd, nCmdShow);
+                UpdateWindow(hWnd);
+            }
+            return hWnd;
+        }
+
+        void SetAcceleratorTable(HACCEL acc) {
+            hAccelTable = acc;
+        }
+
+        AppController* Controller() const {
+            return controller;
+        }
+
+        int RunLoop() {
+            MSG msg;
+
+            // Main message loop:
+            while (GetMessage(&msg, nullptr, 0, 0))
+            {
+                if (auto xamlSourceNative2 = desktopXamlSource.as<IDesktopWindowXamlSourceNative2>()) {
+                    BOOL xamlSourceProcessedMessage = FALSE;
+                    winrt::check_hresult(xamlSourceNative2->PreTranslateMessage(&msg, &xamlSourceProcessedMessage));
+                    if (xamlSourceProcessedMessage) {
+                        continue;
+                    }
+                }
+
+                if (!TranslateAccelerator(msg.hwnd, hAccelTable, &msg))
+                {
+                    TranslateMessage(&msg);
+                    DispatchMessage(&msg);
+                }
+            }
+            return (int)msg.wParam;
+        }
+
+        HWND GetBridgeWindow() const {
+            auto interop = desktopXamlSource.as<IDesktopWindowXamlSourceNative>();
+            HWND hWndXamlIsland = nullptr;
+            winrt::check_hresult(interop->get_WindowHandle(&hWndXamlIsland));
+            return hWndXamlIsland;
+        }
+
+        LRESULT WndProc(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam) {
+            auto interop = desktopXamlSource.as<IDesktopWindowXamlSourceNative>();

```

Comment:
QI/Release() for every message. move to where this is needed to cache the result.